### PR TITLE
Autofix for similar letters

### LINKED ISSUE #54: Add more Fault Tolerance
Would you like to add some fault tolerance such as automatically change `l` to `1`?

### PR DESCRIPTION
**Pull request by LuminousXLB**
fix #54

### CODE REVIEW

File: base58/__init__.py [hunk 1]
```diff
@@ -93,7 +102,7 @@ def b58decode_int(
 
 
 def b58decode(
-    v: Union[str, bytes], alphabet: bytes = BITCOIN_ALPHABET
+    v: Union[str, bytes], alphabet: bytes = BITCOIN_ALPHABET, autofix: bool = False
```
Comment (keis) [inline base58/__init__.py:105]: I think this should be added as a keyword only argument by adding ', *, ' to the signature here and in b56decode_int

File: test_base58.py [hunk 2]
```diff
@@ -66,6 +71,12 @@ def test_check_str():
     assert_that(back, equal_to(b'hello world'))
 
 
+def test_autofix_check_str():
```
Comment (keis) [inline test_base58.py:74]: Could you also add a test with a custom alphabet where the autofix is not applied

File: base58/__init__.py [hunk 3]
```diff
@@ -65,21 +65,30 @@ def b58encode(
...
-def _get_base58_decode_map(alphabet: bytes) -> Mapping[int, int]:
-    map = {char: index for index, char in enumerate(alphabet)}
-    return map
+def _get_base58_decode_map(alphabet: bytes, autofix: bool) -> Mapping[int, int]:
+    invmap = {char: index for index, char in enumerate(alphabet)}
+
+    if autofix:
+        groups = [b'0Oo', b'Il1']
+        for group in groups:
+            pivots = list(filter(lambda x: x in alphabet, group))
```
Comment (keis) [inline base58/__init__.py:74]: A list comprehension is easier to use here and a lookup in the map faster
```suggestion
            pivots = [c for c in alphabet if c in invmap]
```

File: test_base58.py (hunk 2 above)
Comment (LuminousXLB) [reply test_base58.py:74]: fix in 54f46056283a2c5004a8bca8f89d9889d06b3161

File: base58/__init__.py (hunk 1 above)
Comment (LuminousXLB) [reply base58/__init__.py:105]: I'm not sure how to do it. Could you please offer your patch?

File: base58/__init__.py [hunk 4]
```diff
@@ -65,21 +65,30 @@ def b58encode(
...
+            if len(pivots) == 1:
+                for alternative in group:
+                    invmap[alternative] = invmap[pivots[0]]
+
+    return invmap
 
 
 def b58decode_int(
-    v: Union[str, bytes], alphabet: bytes = BITCOIN_ALPHABET
+    v: Union[str, bytes], alphabet: bytes = BITCOIN_ALPHABET, autofix: bool = False
```
Comment (keis) [inline base58/__init__.py:83]:
```suggestion
    v: Union[str, bytes], alphabet: bytes = BITCOIN_ALPHABET, *, autofix: bool = False
```

File: base58/__init__.py [hunk 5]
```diff
@@ -65,21 +65,30 @@ def b58encode(
...
 ) -> int:
     """
     Decode a Base58 encoded string as an integer
     """
     v = v.rstrip()
     v = scrub_input(v)
 
-    map = _get_base58_decode_map(alphabet)
+    map = _get_base58_decode_map(alphabet, autofix)
```
Comment (keis) [inline base58/__init__.py:91]:
```suggestion
    map = _get_base58_decode_map(alphabet, autofix=autofix)
```

File: base58/__init__.py (hunk 1 above)
Comment (keis) [reply base58/__init__.py:105]:
```suggestion
    v: Union[str, bytes], alphabet: bytes = BITCOIN_ALPHABET, *, autofix: bool = False
```

**Comment by LuminousXLB** [comment]
It seems that there's something wrong with `pytest_cov`?

**Comment by keis** [comment]
Huh yeah, looks like a new version was causing some issues. Fixed on master!

Could you rebase and also squash the commits together?

**Comment by LuminousXLB** [comment]
It failed for python 3.5

**Comment by keis** [comment]
ah looks like the mypy type checker is failing in the__main__ module, makes sense considering the functions don't have the same signature anymore.

mypy is only being run  on 3.5 so that's why it's only failing that build

**Comment by keis** [comment]
Typing that bit explicitly solves it

```
    fmap: Dict[Tuple[bool, bool], Callable[[bytes], bytes]] = {
        (False, False): b58encode,
        (False, True): b58encode_check,
        (True, False): b58decode,
        (True, True): b58decode_check
    }
    fun = fmap[(args.decode, args.check)]
```

**Comment by keis** [comment]
Oh that kind of annotation is only 3.6+

hmm, I don't think it's time to drop 3.5 already. I need to think about this one

File: base58/__main__.py [hunk 6]
```diff
@@ -1,8 +1,16 @@
...
 
 from base58 import b58decode, b58decode_check, b58encode, b58encode_check
 
+_fmap: Dict[Tuple[bool, bool], Callable[[bytes], bytes]] = {
+    (False, False): b58encode,
+    (False, True): b58encode_check,
+    (True, False): b58decode,
+    (True, True): b58decode_check
+}
+
```
Comment (keis) [inline base58/__main__.py:13]:
```suggestion
_fmap = {
    (False, False): b58encode,
    (False, True): b58encode_check,
    (True, False): b58decode,
    (True, True): b58decode_check
}  # type: Dict[Tuple[bool, bool], Callable[[bytes], bytes]]

```

Alright, lets just use the old comment style

**Pull request merged**